### DISCUSSION TITLE
Optimistic concurrency across sessions (e.g. when working with ViewModels/DTOs)

### DISCUSSION THREAD

**Discussion by miloszkukla** [ravendb/ravendb#12684] [Q&A] [answered]
Let's consider the an app which make use of view models and talks directly to RavenDb. The classic CRUD scenario would be as follows:

1. LoadProducts:
```
using (var session = documentStore.OpenSession())
{ 
    return session.Query<Product>.Select(x => new ProductViewModel
    {
        Id = x.Id,
        Description = x.Description
    }.ToList();
}
```

2. Modify particular Product via UI (ProductViewModel gets modified)

3. SaveProduct(ProductViewModel productViewModel)
```
using (var session = documentStore.OpenSession())
{ 
    Product product;
    if (productViewModel.Id == null)
    {
        product = new Product();
        session.Store(productViewModel);
        productViewModel.Id = product.Id;
    }
    else
    {
        product = session.Load<Product>(productViewModel.Id) ?? throw new InvalidOperationException();
    }
    product.Description = productViewModel.Description; 
    session.SaveChanges() 
}
```

The similar flow could probably be used if we would replace ViewModels with DTOs in a service.

The question is what would be the recommended way to make use of RavenDb's optimistic concurrency across sessions?

One way that comes to my mind is to keep change vector in ViewModel - I'm not sure however how would RavenDb react if I would do `session.Advanced.GetMetadataFor(product)["@change-vector"] = productViewModel.ChangeVector` or execute `session.Store(product, productViewModel.ChangeVector, product.Id)` for existing product.

**Comment by ayende** [comment, marked as answer]
You need to send the change vector to the client, and then get it back.
On saving it back, you can use the `session.Store(product, productViewModel.ChangeVector, product, product.Id)` overload, yes.
That will instruct RavenDB to do an optimistic concurrency check with the previous value.

do not modify the `@change-vector` metadata, that won't do what you want it to. The overload is the proper and supported way to handle this.
